src/main/java/redis/clients/jedis/JedisClusterCommand.java

```diff
@@ -79,62 +79,87 @@ public T runWithAnyNode() {
     }
   }
 
-  private T runWithRetries(final int slot, int attempts, boolean tryRandomNode, JedisRedirectionException redirect) {
-    if (attempts <= 0) {
-      throw new JedisClusterMaxAttemptsException("No more cluster attempts left.");
-    }
+  private interface ConnectionGetter {
+    Jedis getConnection();
+  }
 
-    Jedis connection = null;
-    try {
+  private T runWithRetries(final int slot) {
+    ConnectionGetter connectionGetter = new ConnectionGetter() {
+      @Override
+      public Jedis getConnection() {
+        return connectionHandler.getConnectionFromSlot(slot);
+      }
+    };
 
-      if (redirect != null) {
-        connection = this.connectionHandler.getConnectionFromNode(redirect.getTargetNode());
-        if (redirect instanceof JedisAskDataException) {
-          // TODO: Pipeline asking with the original command to make it faster....
-          connection.asking();
-        }
-      } else {
-        if (tryRandomNode) {
-          connection = connectionHandler.getConnection();
+    // If we got one redirection, stick with that and don't try anything else
+    ConnectionGetter redirected = null;
+
+    for (int currentAttempt = 0; currentAttempt < this.maxAttempts; currentAttempt++) {
+      Jedis connection = null;
+      try {
+        if (redirected != null) {
+          connection = redirected.getConnection();
         } else {
-          connection = connectionHandler.getConnectionFromSlot(slot);
+          connection = connectionGetter.getConnection();
+        }
+        return execute(connection);
+      } catch (JedisNoReachableClusterNodeException e) {
+        throw e;
+      } catch (JedisConnectionException e) {
+        connectionGetter = handleConnectionProblem(slot, currentAttempt);
+      } catch (JedisRedirectionException e) {
+        redirected = handleRedirection(connection, e);
+      } finally {
+        if (connection != null) {
+          releaseConnection(connection);
         }
```

Comment:
Skip null checking because it is done in `releaseConnection` anyway.

Reply:
Done.